include/aws/crt/Outcome.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/error.h>
+
+#include <utility>
+
+namespace Aws
+{
+    namespace Crt
+    {
+        template <typename R, typename E> // Result, Error
+        class Outcome
+        {
+          public:
+            Outcome() : m_success(false) {}
+
+            Outcome(const R &r) : m_result(r), m_success(true) {}
+
+            Outcome(const E &e) : m_error(e), m_success(false) {}
```

Comment:
Problems for another day: Works until R and E might be confused for one another by the compiler.
Ex: `AwsCrtResult<int> HttpClientStream::GetStatusCode()`

Reply:
True, but I wish we had a specific error type other than int as well, which would make things much less likely.